Replace CasperJS browser tests with Selenium tests.

jmchilton opened the pull request:
See documentation added to run_tests.sh. This can be configured to target local web browsers or Selenium remote services as detailed in that script.

Individual tests can be executed with nosetests directly or groups of tests can share a common test Galaxy spin up when using ./run_tests.sh. If GALAXY_TEST_EXTERNAL=<url> is set - it will be respected and Galaxy will not be spun up (remember the URL needs to be reachable from the Selenium server inside the Docker container).

Every test failure writes a current screenshot of Galaxy to database/test_errors.

I will be honest that the desire to move toward selenium is based solely on failing to get the Casper tests to pass consistently due to bugs causing segfaults in the underlying tools (probably casperjs or phantomjs?). Over the past year I have tried multiple versions of dependencies, etc... and it never works out for me. Likewise Martin has never been able to get the tests to run consistently under Jenkins. I don't think Python or JavaScript is inherently better, this wasn't based on a personal preference about what kind of test I want to write.

Despite this being the primary reason, there are clear benefits to Selenium. It tests the actual web browsers we support, generates screenshots, has better documented and more robust tooling, can scale across clusters. Less clear, but certainly still a benefit of being Python based is that it fits with the rest of the test framework more cleanly than CasperJS.

Finally Carl's last words on CasperJS were "Ditch it".

The following specific tests were added or replaced:

- Implemented tour testing via Selenium (walk the two working stock tours and verify elements available and clicks are valid). This is how I discovered #3206.
- Added completely new workflow GUI test (basic creation from URL and in editor).
- Replaced broken CasperJS registration tests (text and expectations now wrong) with a Selenium variant that works against dev.
- Replaced broken anonymous history CasperJS tests with working Selenium tests.
- Replaced broken upload CasperJS tests with working Selenium tests.
- Replaced history options CasperJS tests with Selenium tests.
- Replaced login CasperJS tests with Selenium tests.
- Replaced history-share-tests.js with test_history_options.py
- Replaced history-panel-tests.js with test_history_panel.py
- Replaced a big part of hda-state-tests.js with test_history_dataset_state.py

I'm confident these utilities represent a sharable and higher-level abstraction around functional testing of Galaxy that can be used outside of Galaxy's testing framework. So a subset of the functional test stuff is in a separate module with minimal dependencies that I intend to make stand alone and pip installable like galaxy-lib.

This module consists of:

- Sizzle stuff in its own package. This code allows Selenium to reason with jQuery selectors instead of vanilla CSS selectors.
- Functionality for creating a Selenium driver and virtual display.
- The ``HasDriver`` mixin - this provides higher level navigation utilities not specifically tied to Galaxy.
- A package with abstractions for navigating Galaxy. This provides a higher-level interfactor for things such as logging in and out, registering a user, navigating menus, fetching Galaxy style tooltips and error messages, and walking Galaxy tours.

In order to demonstrate this new module is useful outside of Galaxy tests - I've included a CLI package to ease building argparse utilities around these abstractions and implemented a simple demonstration script that walks a Galaxy tour and dumps screenshots of Galaxy at each step to a folder. This is more of a demonstration of the utilities than an actual polished end user tool - though it might be helpful in linting and debugging tours.

In the future I hope to extend scripting based on this module to implement

- Periodic deployment testing to ensure things like Jupyter work in production.
- A best practice recipe for Galaxy QA testing by deployers.
- An utility to generate dozens of screenshots of various aspects of Galaxy so we can quickly visually inspect ever part of the GUI before big releases.

xref galaxyproject/starforge#115 xref #1419

Beyond 82728df1555141be4fa3e7be16e151124b1dfc89, this PR contains a few other testing goodies.

 - https://github.com/galaxyproject/galaxy/commit/16b027e12509356a53a4fd1268855d913e5438f3 Is a small adjustment that allows pretty much all the tests to spin up Galaxy as needed - so any of the API or selenium tests can be run directly with nosetests and do not actually need to come through ``run_tests.sh`` anymore for one off executions.
- A fix for GALAXY_TEST_EXTERNAL
- This PR also includes migrations of some CasperJS API tests to the API test framework.
- This PR also contains a set of Jenkins test scripts. Makes it easier to run exact tests (or subset of tests) from Jenkins locally. The Jenkins test script for Selenium provides a dead simple way to run Selenium, Postgres, and Galaxy together with docker-compose. Try it out with:

```
./.ci/jenkins/selenium/run_tests.sh test/selenium_tests/test_anon_history.py
```

I don't have a Mac, but I think this now represents the easier way to try out these Selenium tests on Mac OS X. It should also make getting these tests running in Jenkins easier.

jmchilton committed:
Configure a test Galaxy as needed for functional test cases.

./run_tests.sh is still good for launching one Galaxy for a bunch of different test cases - but now you can do really simple things like:

```
$ nosetests test/api/test_tours.py
```

In this scenario, the test case class will check if run_tests.sh has created a test driver to use and if it hasn't - one will be made just for the test case (like the new integration tests).

jmchilton committed:
Replace CasperJS browser tests with Selenium tests.

See documentation added to run_tests.sh. This can be configured to target local web browsers or Selenium remote services as detailed in that script.

Individual tests can be executed with nosetests directly or groups of tests can share a common test Galaxy spin up when using ./run_tests.sh. If GALAXY_TEST_EXTERNAL=<url> is set - it will be respected and Galaxy will not be spun up (remember the URL needs to be reachable from the Selenium server inside the Docker container).

Every test failure writes a current screenshot of Galaxy to database/test_errors.

I will be honest that the desire to move toward selenium is based solely on failing to get the Casper tests to pass consistently due to bugs causing segfaults in the underlying tools (probably casperjs or phantomjs?). Over the past year I have tried multiple versions of dependencies, etc... and it never works out for me. Likewise Martin has never been able to get the tests to run consistently under Jenkins. I don't think Python or JavaScript is inherently better, this wasn't based on a personal preference about what kind of test I want to write.

Despite this being the primary reason, there are clear benefits to Selenium. It tests the actual web browsers we support, generates screenshots, has better documented and more robust tooling, can scale across clusters. Less clear, but certainly still a benefit of being Python based is that it fits with the rest of the test framework more cleanly than CasperJS.

Finally Carl's last words on CasperJS were "Ditch it".

The following specific tests were added or replaced:

- Implemented tour testing via Selenium (walk the two working stock tours and verify elements avaiable and clicks are valid). This is how I discovered #3206.
- Added completely new workflow GUI test (basic creation from URL and in editor).
- Replaced broken CasperJS registeration tests (text and expectations now wrong) with a Selenium variant that works against dev.
- Replaced broken anonymous history CasperJS tests with working Selenium tests.
- Replaced broken upload CasperJS tests with working Selenium tests.
- Replaced history options CasperJS tests with Selenium tests.
- Replaced login CasperJS tests with Selenium tests.
- Replaced history-share-tests.js with test_history_options.py
- Replaced history-panel-tests.js with test_history_panel.py
- Replaced a big part of hda-state-tests.js with test_history_dataset_state.py

I'm confident these utilities represent a sharable and higher-level abstraction around functional testing of Galaxy that can be used outside of Galaxy's testing framework. So a subset of the functional test stuff is in a separate module with minimal dependencies that I intend to make stand alone and pip installable like galaxy-lib.

This module consists of:

- Sizzle stuff in its own package. This code allows Selenium to reason with jQuery selectors instead of vanilla CSS selectors.
- Functionality for creating a Selenium driver and virtual display.
- The ``HasDriver`` mixin - this provides higher level navigation utilities not specifically tied to Galaxy.
- A package with abstractions for navigating Galaxy. This provides a higher-level interfactor for things such as logging in and out, registering a user, navigating menus, fetching Galaxy style tooltips and error messages, and walking Galaxy tours.

In order to demonstrate this new module is useful outside of Galaxy tests - I've included a CLI package to ease building argparse utilities around these abstractions and implemented a simple demonstration script that walks a Galaxy tour and dumps screenshots of Galaxy at each step to a folder. This is more of a demonstration of the utilities than an actual polished end user tool - though it might be helpful in linting and debugging tours.

In the future I hope to extend this scripting to implement

- Periodic deployment testing to ensure things like Jupyter work in production.
- A best practice recipe for Galaxy QA testing by deployers.
- An utility to generate dozens of screenshots of various aspects of Galaxy so we can quickly visually inspect ever part of the GUI before big releases.

xref https://github.com/galaxyproject/starforge/pull/115 xref #1419

jmchilton committed:
Allow test data resolver to work with GALAXY_TEST_EXTERNAL.

jmchilton committed:
Remove CasperJS api-batch-tests.js - already available in API test framework.

jmchilton committed:
Remove CasperJS API user and history tests...

... more complete testing for both is available in API tests.

jmchilton committed:
Move CasperJS history API tests in api/test_histories.py.

jmchilton committed:
Create simple scripts for all Jenkins tests.

Makes it easier to run exact tests (or subset of tests) from Jenkins locally.

Selenium script is particularily interesting because it represents a very simple entry point into selenium testing.

dannon commented:
@jmchilton Nice, I'll retest this on osx now.

dannon commented:
```
Traceback (most recent call last):
  File "/Users/yoplait/work/galaxy/test/selenium_tests/framework.py", line 87, in setUp
    self.display = driver_factory.virtual_display_if_enabled(headless_selenium())
  File "/Users/yoplait/work/galaxy/test/selenium_tests/framework.py", line 178, in headless_selenium
    return asbool(GALAXY_TEST_SELENIUM_HEADLESS)
  File "/Users/yoplait/work/galaxy/lib/galaxy/util/__init__.py", line 819, in asbool
    raise ValueError("String is not true/false: %r" % obj)
ValueError: String is not true/false: 'auto'
```

Instead of 'auto', should that be true or false by default?

jmchilton committed:
Attempt fix of logic for GALAXY_TEST_SELENIUM_HEADLESS == "auto".

Pointed out by @dannon.

jmchilton commented:
Sorry @dannon and thanks for the report - I was only testing with ``pyvirtualdisplay`` available in my virtualenv. I meant for the default to be "auto", I just didn't implement it correctly at all. I pushed out a fix I think for this - but I will do some more testing after other fires have been put out.

dannon commented:
@jmchilton No rush, I set the env var and continued testing -- looks like it's working pretty nicely to me so far.  I'm excited about the tour walkthrough bit, that's something Sam and I had also wanted to use for testing!

martenson commented:
testing the jenkins deployment, getting stuck on docker-compose (https://jenkins.galaxyproject.org/job/martenson-selenium/2/console)

jmchilton commented:
@martenson https://github.com/docker/compose/issues/1895#issuecomment-132878499

Can you try adding ``export COMPOSE_API_VERSION=auto`` to the Jenkins job? If that works I'll add it to the script in this PR.

martenson commented:
I still have `Waiting on docker-compose managed Galaxy server - http://localhost:46439/.`

```
root@n4:~# docker ps
CONTAINER ID        IMAGE                                       COMMAND                  CREATED             STATUS              PORTS                     NAMES
b5c08da8ccc8        selenium/standalone-chrome:3.0.1-aluminum   "/opt/bin/entry_point"   18 minutes ago      Up 18 minutes       0.0.0.0:43383->4444/tcp   selenium_selenium_1
317f25d2147d        postgres                                    "/docker-entrypoint.s"   42 minutes ago      Up 19 minutes       5432/tcp                  selenium_postgres_1
```

https://jenkins.galaxyproject.org/job/martenson-selenium/6/console

jmchilton commented:
I think there is a failed Galaxy container from a previous step - I should clean out the workspace before I I start a new test. I'll give this another shot...

martenson commented:
@jmchilton I am trying to keep an eye on the containers and cleaning them; also the workspace is always fresh

jmchilton committed:
Update selenium test's use of docker-compose to prevent interference across builds.

jmchilton commented:
@martenson Sorry - I don't think I state what I meant to very clearly. I think fa3d2063c4df595e5d0bfd47d7bf30b8b3a567eb should prevent that error - I kept getting stuff like that. I would kill the container but docker-compose would restart it instead of giving me a fresh one. fa3d2063c4df595e5d0bfd47d7bf30b8b3a567eb should probably fix that.

martenson commented:
I resolved the docker versioning problem, now I am always getting stuck at the docker compose, see the jenkins logs

jmchilton commented:
```
Starting selenium_postgres_1
Recreating selenium_galaxy_1
Recreating selenium_selenium_1
```

Yup - this is a problem I don't know why that last fix didn't work. I'll work on robustness of monitoring and cleaning things up.

jmchilton committed:
Fix and enhance Jenkins test for Selenium.

- If condition on fa3d2063c4df595e5d0bfd47d7bf30b8b3a567eb was exactly wrong.
- Check for docker-compose services up before and during polling to wait for Galaxy - if Galaxy isn't going to work just die.
- Improved logging.

jmchilton commented:
@martenson Sorry this hasn't gone well - I think e731b6f will get us a lot closer and even if there are still problems they should be caught, logged, and cleaned up afterwards now. Hope it helps.

jmchilton committed:
Another fix for Selenium Jenkins.

docker-compose strips out dashes and underscores in the project name when creating containers.

jmchilton commented:
We have passing tests on Jenkins... https://jenkins.galaxyproject.org/job/martenson-selenium/14/. Thanks for all the help on that front @martenson.

jmchilton committed:
Add bed datatype to framework tests because Filter1 now a framework test tool.

nsoranzo commented on scripts/dump_tour.py:
```diff
@@ -0,0 +1,52 @@
+    parser.add_argument('-o', '--output', default="tour_dump_TIMESTAMP", help='directory to dump tour to')
+    parser = cli.add_selenium_arguments(parser)
+    return parser
+
```
An extra blank line is needed here to make flake8 happy.

jmchilton committed:
PEP-8 fix as pointed out by @nsoranzo.

bgruening commented:
This is awesome! Thanks to all!
